modules/seqexec/server/src/main/scala/seqexec/server/gsaoi/GsaoiController.scala

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2016-2019 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package seqexec.server.gsaoi
+
+import cats.Applicative
+import cats.Show
+import cats.implicits._
+import seqexec.model.dhs.ImageFileId
+import seqexec.server.gsaoi.GsaoiController.DCConfig
+import seqexec.server.gsaoi.GsaoiController.GsaoiConfig
+import seqexec.server.ObserveCommand
+import seqexec.server.Progress
+import squants.Time
+import squants.time.TimeConversions._
+import shapeless.tag.@@
+
+trait GsaoiController[F[_]] {
+
+  def applyConfig(config: GsaoiConfig): F[Unit]
+
+  def observe(fileId: ImageFileId, cfg: DCConfig): F[ObserveCommand.Result]
+
+  def endObserve: F[Unit]
+
+  def stopObserve: F[Unit]
+
+  def abortObserve: F[Unit]
+
+  def observeProgress(total: Time): fs2.Stream[F, Progress]
+
+  def calcTotalExposureTime(cfg: DCConfig)(
+    implicit ev:                 Applicative[F]): F[Time] = {
+    val readFactor  = 2.2
+    val readOutTime = 300
+
+    (cfg.coadds * cfg.exposureTime * readFactor + readOutTime.seconds).pure[F]
+  }
+
+}
+
+trait CoaddsI
+trait NumberOfFowSamplesI
+
+sealed trait WindowCover extends Product with Serializable
+
+object WindowCover {
+  case object Closed extends WindowCover
+  case object Opened extends WindowCover
+}
+
+object GsaoiController {
+  // DC
+  type ReadMode           = edu.gemini.spModel.gemini.gsaoi.Gsaoi.ReadMode
+  type Roi                = edu.gemini.spModel.gemini.gsaoi.Gsaoi.Roi
+  type Coadds             = Int @@ CoaddsI
+  type ExposureTime       = Time
+  type NumberOfFowSamples = Int @@ NumberOfFowSamplesI
+
+  // CC
+  type Filter       = edu.gemini.spModel.gemini.gsaoi.Gsaoi.Filter
+  type OdgwSize     = edu.gemini.spModel.gemini.gsaoi.Gsaoi.OdgwSize
```

Comment:
This one will be tricky, because it can be overridden from TCC.

Reply:
I suppose we still have to read it from the sequence and override it on the controller?

Reply:
It will probably be included in the GuideConfig stored in GuideConfigDb. Then, at the configuration time, the final value has to be calculated